tasks/indexer/chainheadindexer.go

```diff
@@ -0,0 +1,107 @@
+package indexer
+
+import (
+	"context"
+
+	logging "github.com/ipfs/go-log/v2"
+	"go.opentelemetry.io/otel/api/global"
+	"golang.org/x/xerrors"
+
+	lotus_api "github.com/filecoin-project/lotus/api"
+	store "github.com/filecoin-project/lotus/chain/store"
+
+	"github.com/filecoin-project/sentinel-visor/lens"
+	"github.com/filecoin-project/sentinel-visor/storage"
+)
+
+var log = logging.Logger("indexer")
+
+// NewChainHeadIndexer creates a new ChainHeadIndexer. confidence sets the number of tipsets that will be held
+// in a cache awaiting possible reversion. Tipsets will be written to the database when they are evicted from
+// the cache due to incoming later tipsets.
+func NewChainHeadIndexer(d *storage.Database, node lens.API, confidence int) *ChainHeadIndexer {
+	return &ChainHeadIndexer{
+		node:       node,
+		storage:    d,
+		confidence: confidence,
+		cache:      NewTipSetCache(confidence),
+	}
+}
+
+// ChainHeadIndexer is a task that indexes blocks by following the chain head.
+type ChainHeadIndexer struct {
+	node       lens.API
+	storage    *storage.Database
+	confidence int          // size of tipset cache
+	cache      *TipSetCache // caches tipsets for possible reversion
+}
+
+// Run starts following the chain head and blocks until the context is done or
+// an error occurs.
+func (c *ChainHeadIndexer) Run(ctx context.Context) error {
+	hc, err := c.node.ChainNotify(ctx)
+	if err != nil {
+		return xerrors.Errorf("chain notify: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case headEvents, ok := <-hc:
+			if !ok {
+				log.Warn("ChainNotify channel closed, stopping Indexer")
+				return nil
+			}
+			if err := c.index(ctx, headEvents); err != nil {
+				return xerrors.Errorf("index: %w", err)
+			}
+		}
+	}
+}
+
+func (c *ChainHeadIndexer) index(ctx context.Context, headEvents []*lotus_api.HeadChange) error {
+	ctx, span := global.Tracer("").Start(ctx, "ChainHeadIndexer.index")
+	defer span.End()
+
+	data := NewUnindexedBlockData()
+
+	var height int64
+	for _, ch := range headEvents {
+		switch ch.Type {
+		case store.HCCurrent:
+			fallthrough
+		case store.HCApply:
+			log.Debugw("add tipset", "event", ch.Type, "height", ch.Val.Height(), "tipset", ch.Val.Key().String())
+			if int64(ch.Val.Height()) > height {
+				height = int64(ch.Val.Height())
+			}
+			tail, err := c.cache.Add(ch.Val)
+			if err != nil {
+				log.Errorw("tipset cache", "error", err.Error())
```

Comment:
So this keeps logging errors when the new value height is the same as the current head height of the cache. I modified the `Add` function to return the current head with this error so I could see what the diff between the heights is and this is the logs that kept repeating:
```
2020-10-03T12:25:30.672+1000	ERROR	indexer	indexer/chainheadindexer.go:93	tipset cache	{"error": "reverted tipset does not match current head"}
2020-10-03T12:25:30.672+1000	ERROR	indexer	indexer/chainheadindexer.go:81	tipset cache	{"error": "added tipset height lower than current head", "oldheight": "112851", "newheight": "112851"}
2020-10-03T12:25:31.126+1000	ERROR	indexer	indexer/chainheadindexer.go:93	tipset cache	{"error": "reverted tipset does not match current head"}
2020-10-03T12:25:31.126+1000	ERROR	indexer	indexer/chainheadindexer.go:81	tipset cache	{"error": "added tipset height lower than current head", "oldheight": "112851", "newheight": "112851"}
2020-10-03T12:25:31.128+1000	ERROR	indexer	indexer/chainheadindexer.go:93	tipset cache	{"error": "reverted tipset does not match current head"}
2020-10-03T12:25:31.128+1000	ERROR	indexer	indexer/chainheadindexer.go:81	tipset cache	{"error": "added tipset height lower than current head", "oldheight": "112851", "newheight": "112851"}
2020-10-03T12:25:31.130+1000	ERROR	indexer	indexer/chainheadindexer.go:93	tipset cache	{"error": "reverted tipset does not match current head"}
2020-10-03T12:25:31.130+1000	ERROR	indexer	indexer/chainheadindexer.go:81	tipset cache	{"error": "added tipset height lower than current head", "oldheight": "112851", "newheight": "112851"}
2020-10-03T12:25:37.094+1000	ERROR	indexer	indexer/chainheadindexer.go:93	tipset cache	{"error": "reverted tipset does not match current head"}
2020-10-03T12:25:37.094+1000	ERROR	indexer	indexer/chainheadindexer.go:81	tipset cache	{"error": "added tipset height lower than current head", "oldheight": "112851", "newheight": "112851"}
```

If we attempted to add an actual old height, I would want to know about it but at the same height is going to happen.

Reply:
Possibly because of the fallthrough for HCCurrent events. I don't have a fully synced node to try this against yet.

Reply:
Added explicit handling for current head

Reply:
Seems to be resolved. :)